### ISSUE TITLE
Training not converging with flash-attention enabled

### ISSUE THREAD

**Issue by toby-clark4** [Dao-AILab/flash-attention#1597]
I've tried to implement attention in a Hugging Face-compatible model using the code below. However, when I train my model, train loss plateaus very quickly at a much higher value than when I fall back to standard attention. I feel like this is likely an issue with my code but wanted to check just in case there is an issue I'm missing. 
```
query_layer = self.transpose_for_scores(self.query(hidden_states))
key_layer = self.transpose_for_scores(self.key(hidden_states))
value_layer = self.transpose_for_scores(self.value(hidden_states))

if sinusoidal_pos is not None:
    if self.rotary_value:
        query_layer, key_layer, value_layer = self.apply_rotary_position_embeddings(
            sinusoidal_pos, query_layer, key_layer, value_layer
        )
    else:
        query_layer, key_layer = self.apply_rotary_position_embeddings(
            sinusoidal_pos, query_layer, key_layer
        )

# permute to (bsz, seqlen, num_heads, head_dim)
query_layer = query_layer.permute(0, 2, 1, 3).contiguous()
key_layer = key_layer.permute(0, 2, 1, 3).contiguous()
value_layer = value_layer.permute(0, 2, 1, 3).contiguous()            

# unpad each tensor to skip padding tokens
attention_mask = attention_mask.squeeze(1).squeeze(1).bool()
q_unpad, indices, cu_seqlens_q, max_seqlen_q, _ = unpad_input(query_layer, attention_mask.bool())
k_unpad, _, cu_seqlens_k, max_seqlen_k, _ = unpad_input(key_layer, attention_mask.bool())
v_unpad, _, _, _, _ = unpad_input(value_layer, attention_mask.bool())

q_unpad = q_unpad.to(torch.bfloat16)
k_unpad = k_unpad.to(torch.bfloat16)
v_unpad = v_unpad.to(torch.bfloat16)

context_unpad = flash_attn_varlen_func(
    q_unpad, k_unpad, v_unpad,
    cu_seqlens_q=cu_seqlens_q,
    cu_seqlens_k=cu_seqlens_k,
    max_seqlen_q=max_seqlen_q,
    max_seqlen_k=max_seqlen_k,
    dropout_p=self.dropout.p,
    causal=self.is_decoder,
)

# restore padding (bsz, seq_len, num_heads, head_dim)
context = pad_input(context_unpad, indices, bsz, seqlen)

# reshape to (bsz, seqlen, num_heads * head_dim)
context = context.reshape(bsz, seqlen, self.all_head_size)

if output_attentions:
    return (context, None)
return (context,)
```
